// @tatumio/shared-core/src/lib/derivation-path.common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.COMMON_TESTNET_DERIVATION_PATH = exports.DERIVATION_PATH = exports.getDerivationPath = void 0;
const Blockchain_1 = require("./models/Blockchain");
const getDerivationPath = (blockchain, options) => {
    return (options === null || options === void 0 ? void 0 : options.testnet) ? TESTNET_DERIVATION_PATH[blockchain] : exports.DERIVATION_PATH[blockchain];
};
exports.getDerivationPath = getDerivationPath;
exports.DERIVATION_PATH = {
    BTC: "m/44'/0'/0'/0",
    LTC: "m/44'/2'/0'/0",
    ETH: "m/44'/60'/0'/0",
    DOGE: "m/44'/3'/0'/0",
    CELO: "m/44'/52752'/0'/0",
    POLYGON: "m/44'/966'/0'/0",
    KCS: "m/44'/60'/0'/0",
    HARMONY: "m/44'/1023'/0'/0",
    KLAY: "m/44'/8217'/0'/0",
    BSC: "m/44'/60'/0'/0",
    BCH: "m/44'/145'/0'/0",
    TRON: "m/44'/195'/0'/0",
    EGLD: "m/44'/508'/0'/0'",
    // TBD
    ALGO: '@TODO - TBD',
    ADA: "m/1852'/1815'/0'",
    FLOW: "m/44'/539'/0'/0",
    NEO: "m/44'/888'/0'/0",
    SOL: "m/44'/501'/0'/0'",
    VET: "m/44'/818'/0'/0",
    XDC: "m/44'/550'/0'/0",
    XLM: "m/44'/148'/0'",
    XRP: "m/44'/144'/0'/0",
    TEZOS: "m/44'/1729'/0'/0",
    EON: "m/44'/60'/0'/0",
    CHILIZ: "m/44'/2182'/0'/0",
    FLR: "m/44'/60'/0'/0",
    CRO: "m/44'/60'/0'/0",
    BASE: "m/44'/60'/0'/0",
    AVAX: "m/44'/60'/0'/0",
    OPTIMISM: "m/44'/60'/0'/0",
    FTM: "m/44'/60'/0'/0",
    TON: "m/44'/60'/0'/0",
    ZK_SYNC: "m/44'/60'/0'/0",
    RON: "m/44'/60'/0'/0",
    S: "m/44'/60'/0'/0",
    KAIA: "m/44'/60'/0'/0",
    BERA: "m/44'/60'/0'/0",
    SUI: "m/44'/60'/0'/0",
};
exports.COMMON_TESTNET_DERIVATION_PATH = "m/44'/1'/0'/0";
const TESTNET_DERIVATION_PATH_MAPPING = Object.values(Blockchain_1.Blockchain)
    .map((v) => {
    return { [v]: exports.COMMON_TESTNET_DERIVATION_PATH };
})
    .reduce((obj, item) => (Object.assign(Object.assign({}, obj), item)), {});
const TESTNET_DERIVATION_PATH = Object.assign(Object.assign({}, TESTNET_DERIVATION_PATH_MAPPING), { 
    /**
     * BCH historically uses mainnet derivation path.
     * It's not right, but changing it will break production users experience
     */
    BCH: exports.DERIVATION_PATH.BCH });
