// @tatumio/api-client/src/lib/models/Currency.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NFT_SUPPORTED_CURRENCIES = exports.BASE_BASED_CURRENCIES = exports.BASE_CURRENCIES = exports.OPTIMISM_BASED_CURRENCIES = exports.OPTIMISM_CURRENCIES = exports.ARB_BASED_CURRENCIES = exports.ARB_CURRENCIES = exports.BSC_BASED_CURRENCIES = exports.MATIC_BASED_CURRENCIES = exports.ETH_BASED_CURRENCIES = exports.MATIC20_CURRENCIES = exports.CELO_CURRENCIES = exports.FLOW_CURRENCIES = exports.TRON_CURRENCIES = exports.NATIVE_CURRENCIES = exports.BEP20_CURRENCIES = exports.ERC20_CURRENCIES = exports.Currency = void 0;
var Currency;
(function (Currency) {
    Currency["BTC"] = "BTC";
    Currency["AVAX"] = "AVAX";
    Currency["FTM"] = "FTM";
    Currency["REVV"] = "REVV";
    Currency["SAND"] = "SAND";
    Currency["ADA"] = "ADA";
    Currency["BCH"] = "BCH";
    Currency["LTC"] = "LTC";
    Currency["ONE"] = "ONE";
    Currency["ETH"] = "ETH";
    Currency["ETC"] = "ETC";
    Currency["EGLD"] = "EGLD";
    Currency["MATIC"] = "MATIC";
    Currency["GAMEE"] = "GAMEE";
    Currency["MATIC_ETH"] = "MATIC_ETH";
    Currency["POL_ETH"] = "POL_ETH";
    Currency["USDC_MATIC"] = "USDC_MATIC";
    Currency["USDC_MATIC_NATIVE"] = "USDC_MATIC_NATIVE";
    Currency["USDC_BSC"] = "USDC_BSC";
    Currency["USDT_BSC"] = "USDT_BSC";
    Currency["USDC_SOL"] = "USDC_SOL";
    Currency["RMD"] = "RMD";
    Currency["GMC"] = "GMC";
    Currency["GMC_BSC"] = "GMC_BSC";
    Currency["CELO"] = "CELO";
    Currency["COIIN"] = "COIIN";
    Currency["COIIN_BSC"] = "COIIN_BSC";
    Currency["CUSD"] = "CUSD";
    Currency["CEUR"] = "CEUR";
    Currency["FLOW"] = "FLOW";
    Currency["FUSD"] = "FUSD";
    Currency["BSC"] = "BSC";
    Currency["XDC"] = "XDC";
    Currency["DOGE"] = "DOGE";
    Currency["XRP"] = "XRP";
    Currency["XLM"] = "XLM";
    Currency["USDT"] = "USDT";
    Currency["TRON"] = "TRON";
    Currency["LEO"] = "LEO";
    Currency["LINK"] = "LINK";
    Currency["LISK"] = "LISK";
    Currency["FREE"] = "FREE";
    Currency["MKR"] = "MKR";
    Currency["USDC"] = "USDC";
    Currency["UNI"] = "UNI";
    Currency["BAT"] = "BAT";
    Currency["TUSD"] = "TUSD";
    Currency["PAX"] = "PAX";
    Currency["PLTC"] = "PLTC";
    Currency["XCON"] = "XCON";
    Currency["MMY"] = "MMY";
    Currency["PAXG"] = "PAXG";
    Currency["VET"] = "VET";
    Currency["HAG"] = "HAG";
    Currency["BETH"] = "BETH";
    Currency["BUSD"] = "BUSD";
    Currency["BBTC"] = "BBTC";
    Currency["BADA"] = "BADA";
    Currency["WBNB"] = "WBNB";
    Currency["BDOT"] = "BDOT";
    Currency["BXRP"] = "BXRP";
    Currency["BLTC"] = "BLTC";
    Currency["BBCH"] = "BBCH";
    Currency["CAKE"] = "CAKE";
    Currency["BUSD_BSC"] = "BUSD_BSC";
    Currency["B2U_BSC"] = "B2U_BSC";
    Currency["WBTC"] = "WBTC";
    Currency["USDT_TRON"] = "USDT_TRON";
    Currency["USDT_MATIC"] = "USDT_MATIC";
    Currency["LATOKEN"] = "LATOKEN";
    Currency["INRT_TRON"] = "INRT_TRON";
    Currency["ALGO"] = "ALGO";
    Currency["SOL"] = "SOL";
    Currency["KCS"] = "KCS";
    Currency["NEO"] = "NEO";
    /**
     * @deprecated Use `KAIA` instead
     */
    Currency["KLAY"] = "KLAY";
    Currency["EOS"] = "EOS";
    Currency["ETH_ARB"] = "ETH_ARB";
    Currency["ETH_OP"] = "ETH_OP";
    Currency["NEAR"] = "NEAR";
    Currency["CRO"] = "CRO";
    Currency["RSK"] = "RSK";
    Currency["AURORA"] = "AURORA";
    Currency["GNO"] = "GNO";
    Currency["DOT"] = "DOT";
    Currency["KSM"] = "KSM";
    Currency["OASIS"] = "OASIS";
    Currency["TEZOS"] = "TEZOS";
    Currency["PALM"] = "PALM";
    Currency["GLMR"] = "GLMR";
    Currency["INTENT"] = "INTENT";
    Currency["EURTENT"] = "EURTENT";
    Currency["GOLDAX"] = "GOLDAX";
    Currency["ZCASH"] = "ZCASH";
    Currency["ZEC"] = "ZEC";
    Currency["ZIL"] = "ZIL";
    Currency["ZEN"] = "ZEN";
    Currency["CHZ"] = "CHZ";
    Currency["ISLM"] = "ISLM";
    Currency["FLR"] = "FLR";
    Currency["ETH_BASE"] = "ETH_BASE";
    Currency["KADENA"] = "KADENA";
    Currency["ROSTRUM"] = "ROSTRUM";
    Currency["ATOM"] = "ATOM";
    Currency["IOTA"] = "IOTA";
    Currency["ELECTRUM"] = "ELECTRUM";
    Currency["CSPR"] = "CSPR";
    Currency["TON"] = "TON";
    Currency["ZK_SYNC"] = "ZK_SYNC";
    Currency["USDC_ARB"] = "USDC_ARB";
    Currency["USDC_OP"] = "USDC_OP";
    Currency["USDC_BASE"] = "USDC_BASE";
    Currency["USDT_ARB"] = "USDT_ARB";
    Currency["USDT_OP"] = "USDT_OP";
    Currency["USDT_BASE"] = "USDT_BASE";
    Currency["RON"] = "RON";
    Currency["S"] = "S";
    Currency["KAIA"] = "KAIA";
    Currency["BERA"] = "BERA";
    Currency["SUI"] = "SUI";
})(Currency = exports.Currency || (exports.Currency = {}));
exports.ERC20_CURRENCIES = [
    Currency.USDT,
    Currency.LEO,
    Currency.LINK,
    Currency.UNI,
    Currency.MATIC_ETH,
    Currency.POL_ETH,
    Currency.BUSD,
    Currency.SAND,
    Currency.REVV,
    Currency.LATOKEN,
    Currency.COIIN,
    Currency.FREE,
    Currency.XCON,
    Currency.MKR,
    Currency.USDC,
    Currency.BAT,
    Currency.GMC,
    Currency.TUSD,
    Currency.PAX,
    Currency.PLTC,
    Currency.MMY,
    Currency.PAXG,
    Currency.WBTC,
];
exports.BEP20_CURRENCIES = [
    Currency.BETH,
    Currency.BBTC,
    Currency.RMD,
    Currency.USDC_BSC,
    Currency.USDT_BSC,
    Currency.COIIN_BSC,
    Currency.B2U_BSC,
    Currency.BADA,
    Currency.WBNB,
    Currency.GMC_BSC,
    Currency.BDOT,
    Currency.BXRP,
    Currency.BLTC,
    Currency.BBCH,
    Currency.HAG,
    Currency.CAKE,
    Currency.BUSD_BSC,
];
exports.NATIVE_CURRENCIES = [
    Currency.BTC,
    Currency.ONE,
    Currency.ADA,
    Currency.FLOW,
    Currency.BSC,
    Currency.XDC,
    Currency.LTC,
    Currency.DOGE,
    Currency.TRON,
    Currency.BCH,
    Currency.ETH,
    Currency.XRP,
    Currency.XLM,
    Currency.CELO,
    Currency.MATIC,
    Currency.VET,
    Currency.EGLD,
    Currency.ALGO,
    Currency.SOL,
    Currency.KCS,
    Currency.KLAY,
    Currency.ZK_SYNC,
];
exports.TRON_CURRENCIES = [Currency.TRON, Currency.USDT_TRON, Currency.INRT_TRON];
exports.FLOW_CURRENCIES = [Currency.FLOW, Currency.FUSD];
exports.CELO_CURRENCIES = [Currency.CELO, Currency.CEUR, Currency.CUSD];
exports.MATIC20_CURRENCIES = [
    Currency.USDC_MATIC,
    Currency.USDC_MATIC_NATIVE,
    Currency.GAMEE,
    Currency.USDT_MATIC,
    Currency.INTENT,
    Currency.EURTENT,
    Currency.GOLDAX,
];
exports.ETH_BASED_CURRENCIES = [Currency.ETH, ...exports.ERC20_CURRENCIES];
exports.MATIC_BASED_CURRENCIES = [Currency.MATIC, ...exports.MATIC20_CURRENCIES];
exports.BSC_BASED_CURRENCIES = [Currency.BSC, ...exports.BEP20_CURRENCIES];
exports.ARB_CURRENCIES = [Currency.USDC_ARB, Currency.USDT_ARB];
exports.ARB_BASED_CURRENCIES = [Currency.ETH_ARB, ...exports.ARB_CURRENCIES];
exports.OPTIMISM_CURRENCIES = [Currency.USDC_OP, Currency.USDT_OP];
exports.OPTIMISM_BASED_CURRENCIES = [Currency.ETH_OP, ...exports.OPTIMISM_CURRENCIES];
exports.BASE_CURRENCIES = [Currency.USDC_BASE, Currency.USDT_BASE];
exports.BASE_BASED_CURRENCIES = [Currency.ETH_BASE, ...exports.BASE_CURRENCIES];
exports.NFT_SUPPORTED_CURRENCIES = [
    Currency.ETH,
    Currency.MATIC,
    Currency.KCS,
    Currency.ONE,
    Currency.CELO,
    Currency.TRON,
    Currency.FLOW,
    Currency.BSC,
    Currency.KLAY,
];
