// @tatumio/shared-core/src/lib/models/Blockchain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BtcBasedBlockchains = exports.EvmBasedBlockchains = exports.Blockchain = void 0;
var Blockchain;
(function (Blockchain) {
    Blockchain["BTC"] = "BTC";
    Blockchain["BCH"] = "BCH";
    Blockchain["LTC"] = "LTC";
    Blockchain["CELO"] = "CELO";
    Blockchain["SOL"] = "SOL";
    Blockchain["HARMONY"] = "HARMONY";
    Blockchain["ETH"] = "ETH";
    Blockchain["XRP"] = "XRP";
    Blockchain["XLM"] = "XLM";
    Blockchain["DOGE"] = "DOGE";
    Blockchain["VET"] = "VET";
    Blockchain["NEO"] = "NEO";
    Blockchain["BSC"] = "BSC";
    Blockchain["POLYGON"] = "POLYGON";
    Blockchain["FLOW"] = "FLOW";
    Blockchain["TRON"] = "TRON";
    Blockchain["EGLD"] = "EGLD";
    Blockchain["XDC"] = "XDC";
    Blockchain["ADA"] = "ADA";
    Blockchain["ALGO"] = "ALGO";
    Blockchain["KCS"] = "KCS";
    /**
     * @deprecated Use `KAIA` instead
     */
    Blockchain["KLAY"] = "KLAY";
    Blockchain["TEZOS"] = "TEZOS";
    Blockchain["EON"] = "EON";
    Blockchain["CHILIZ"] = "CHILIZ";
    Blockchain["FLR"] = "FLR";
    Blockchain["CRO"] = "CRO";
    Blockchain["BASE"] = "BASE";
    Blockchain["AVAX"] = "AVAX";
    Blockchain["OPTIMISM"] = "OPTIMISM";
    Blockchain["FTM"] = "FTM";
    Blockchain["TON"] = "TON";
    Blockchain["ZK_SYNC"] = "ZK_SYNC";
    Blockchain["RON"] = "RON";
    Blockchain["S"] = "S";
    Blockchain["KAIA"] = "KAIA";
    Blockchain["BERA"] = "BERA";
    Blockchain["SUI"] = "SUI";
})(Blockchain = exports.Blockchain || (exports.Blockchain = {}));
exports.EvmBasedBlockchains = [
    Blockchain.ETH,
    Blockchain.CELO,
    Blockchain.BSC,
    Blockchain.HARMONY,
    Blockchain.POLYGON,
    Blockchain.KCS,
    Blockchain.EON,
    Blockchain.CHILIZ,
    Blockchain.CRO,
    Blockchain.BASE,
    Blockchain.AVAX,
    Blockchain.OPTIMISM,
    Blockchain.FTM,
    Blockchain.ZK_SYNC,
    Blockchain.RON,
    Blockchain.BERA,
];
// @TODO tmp solution
exports.BtcBasedBlockchains = [Blockchain.BTC, Blockchain.LTC, Blockchain.DOGE, Blockchain.BCH];
