// @tatumio/api-client/src/generated/services/ExchangeRateService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExchangeRateService = void 0;
const request_1 = require("../core/request");
class ExchangeRateService {
    /**
     * Get the current exchange rate for exchanging fiat/crypto assets
     * <p><b>1 credit per API call</b></p>
     * <p>Get the current exchange rate for exchanging fiat/crypto assets.</p>
     * <p>By default, the base pair (the target asset) is EUR. When obtaining the exchange rate for an asset (for example, BTC), the value returned by the API expresses the amount of EUR that can be currently exchanged into 1 BTC.</p>
     *
     * @param currency The fiat or crypto asset to exchange
     * @param basePair The target fiat asset to get the exchange rate for
     * @returns ExchangeRate OK
     * @throws ApiError
     */
    static getExchangeRate(currency, basePair = 'EUR') {
        return (0, request_1.request)({
            method: 'GET',
            path: `/v3/tatum/rate/${currency}`,
            query: {
                'basePair': basePair,
            },
            errors: {
                401: `Unauthorized. Not valid or inactive subscription key present in the HTTP Header.`,
                403: `Forbidden. The request is authenticated, but it is not possible to required perform operation due to logical error or invalid permissions.`,
                500: `Internal server error. There was an error on the server during the processing of the request.`,
            },
        });
    }
    /**
     * Get the current exchange rates for exchanging fiat/crypto assets
     * <p><b>1 credit per pair per API call</b></p>
     * <p>Get the current exchange rates for exchanging fiat/crypto assets.</p>
     * <p>When obtaining the exchange rate for an asset (for example, BTC), the value returned by the API expresses the amount of EUR that can be currently exchanged into 1 BTC.</p>
     *
     * @param requestBody
     * @returns ExchangeRate OK
     * @throws ApiError
     */
    static getExchangeRates(requestBody) {
        return (0, request_1.request)({
            method: 'POST',
            path: `/v3/tatum/rate`,
            body: requestBody,
            mediaType: 'application/json',
            errors: {
                401: `Unauthorized. Not valid or inactive subscription key present in the HTTP Header.`,
                403: `Forbidden. The request is authenticated, but it is not possible to required perform operation due to logical error or invalid permissions.`,
                500: `Internal server error. There was an error on the server during the processing of the request.`,
            },
        });
    }
}
exports.ExchangeRateService = ExchangeRateService;
