// @tatumio/shared-blockchain-evm-based/src/lib/evm-based.sdk.errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EvmBasedErrorCodesFromNode = exports.EvmBasedSdkError = void 0;
const shared_abstract_sdk_1 = require("@tatumio/shared-abstract-sdk");
class EvmBasedSdkError extends shared_abstract_sdk_1.SdkError {
    constructor({ error, code }) {
        super({
            originalError: error,
            originalErrorAsString: error === null || error === void 0 ? void 0 : error.name,
            code,
        });
    }
}
exports.EvmBasedSdkError = EvmBasedSdkError;
exports.EvmBasedErrorCodesFromNode = {
    GAS_REQUIRED_EXCEEDS_ALLOWANCE: {
        substring: 'gas required exceeds allowance',
        error({ e, address }) {
            const errorMsg = address
                ? `Insufficient funds to send transaction from account ${address}, ${e}`
                : `Insufficient funds to send transaction, ${e}`;
            return new EvmBasedSdkError({
                code: shared_abstract_sdk_1.SdkErrorCode.EVM_TRANSACTION_ERROR,
                error: new Error(errorMsg),
            });
        },
    },
};
