// @tatumio/shared-testing-common/src/lib/jest.extensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toThrowErrorWithMessageThatIncludes = exports.toThrowSdkErrorWithCode = exports.createResult = void 0;
const createResult = ({ pass, message, notMessage, }) => ({
    message: () => (pass() ? notMessage() : message()).trim(),
    pass: pass(),
});
exports.createResult = createResult;
const toThrowSdkErrorWithCode = (value, code) => (0, exports.createResult)({
    message: () => `expected ${value} to have error code ${code}`,
    notMessage: () => `expected ${value} not to have error code ${code}`,
    pass: () => {
        let errorCode;
        try {
            errorCode = JSON.parse(value.message).errorCode;
        }
        catch (e) {
            return false;
        }
        return errorCode === code;
    },
});
exports.toThrowSdkErrorWithCode = toThrowSdkErrorWithCode;
const toThrowErrorWithMessageThatIncludes = (value, message) => (0, exports.createResult)({
    message: () => `expected ${value} to have error message ${message}`,
    notMessage: () => `expected ${value} not to have error message ${message}`,
    pass: () => {
        let error;
        try {
            error = JSON.parse(JSON.stringify(value, Object.getOwnPropertyNames(value)));
        }
        catch (e) {
            return false;
        }
        return error.message.includes(message);
    },
});
exports.toThrowErrorWithMessageThatIncludes = toThrowErrorWithMessageThatIncludes;
