// @tatumio/shared-abstract-sdk/src/lib/services/virtualAccount.abstract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.abstractSdkVirtualAccount = void 0;
const api_client_1 = require("@tatumio/api-client");
const abstractSdkVirtualAccount = () => {
    return {
        depositAddress: {
            create: api_client_1.BlockchainAddressesService.generateDepositAddress,
            createMultiple: api_client_1.BlockchainAddressesService.generateDepositAddressesBatch,
            checkExists: api_client_1.BlockchainAddressesService.addressExists,
            assign: api_client_1.BlockchainAddressesService.assignAddress,
            remove: api_client_1.BlockchainAddressesService.removeAddress,
            getByAccount: api_client_1.BlockchainAddressesService.getAllDepositAddresses,
        },
        withdrawal: {
            getAll: api_client_1.WithdrawalService.getWithdrawals,
            broadcast: api_client_1.WithdrawalService.broadcastBlockchainTransaction,
            create: api_client_1.WithdrawalService.storeWithdrawal,
            complete: api_client_1.WithdrawalService.completeWithdrawal,
        },
        storeTokenAddress: api_client_1.BlockchainOperationsService.storeTokenAddress,
    };
};
exports.abstractSdkVirtualAccount = abstractSdkVirtualAccount;
