// @tatumio/shared-blockchain-evm-based/src/lib/services/evm-based.smartContract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.evmBasedSmartContract = void 0;
const tslib_1 = require("tslib");
const contracts_1 = require("../contracts");
const smartContract_1 = require("../transactions/smartContract");
const buildSmartContractMethodInvocation = (body, params, methodName, abi) => {
    return {
        fee: body.fee,
        nonce: body.nonce,
        fromPrivateKey: body.fromPrivateKey,
        signatureId: body.signatureId,
        index: body.index,
        amount: body.amount,
        contractAddress: body.contractAddress,
        methodName: methodName,
        params: params,
        methodABI: abi.find((a) => a.name === methodName),
    };
};
const evmBasedSmartContract = (web3) => {
    return {
        helperPrepareSCCall: (body, methodName, params, provider, abi = contracts_1.ListingSmartContract.abi, testnet) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
            const r = buildSmartContractMethodInvocation(body, params, methodName, abi);
            return yield (0, smartContract_1.smartContractWriteMethodInvocation)({ body: r, web3, provider, chain: body.chain });
        }),
    };
};
exports.evmBasedSmartContract = evmBasedSmartContract;
